### PR TITLE
Initial migration from request to got for http-request node

### PR DESCRIPTION
**Pull request by knolleary**
## Proposed changes

This migrates the HTTP Request node to use the `got` module in place of the deprecated `request` module - #2481 

I have got to the point where the unit tests pass **except** for:

 - Proxy handling. I have implemented proxy handling as per the got docs, but I can't get it to work yet. I have raised https://github.com/delvedor/hpagent/issues/19 to try to make some progress. But may need to look at an alternative approach.
 - keepAlive - we don't have unit tests for this... but I know I haven't implemented it yet. It is tied up with the proxy handling somewhat as they both require replacing the HTTP Agent used by `got`.
 
I suspect there will be other edge cases where things have changed - particularly around how errors are reported. But this is a good start.

### CODE REVIEW

File: packages/node_modules/@node-red/nodes/core/network/21-httprequest.js [hunk 1]
```diff
@@ -311,79 +336,96 @@ module.exports = function(RED) {
...
-                        // Calculate request time
-                        var diff = process.hrtime(preRequestTimestamp);
-                        var ms = diff[0] * 1e3 + diff[1] * 1e-6;
-                        var metricRequestDurationMillis = ms.toFixed(3);
-                        node.metric("duration.millis", msg, metricRequestDurationMillis);
-                        if (res.client && res.client.bytesRead) {
-                            node.metric("size.bytes", msg, res.client.bytesRead);
-                        }
+            got(url,opts).then(res => {
+                msg.statusCode = res.statusCode;
```
Comment (kazuhitoyokoi) [inline packages/node_modules/@node-red/nodes/core/network/21-httprequest.js:376]: As I think that @k-toumura said before, we would like to need the starting and ending time of the http request when we investigate logs of Node-RED flow to call REST APIs. We found that the `got` module is suitable to fill the requirement. Can we add the code like the following to output trace log?

```
log.trace('http request: nodeid=' + this.id + ' _msgid=' + msg._msgid
        + ' url=' + res.url + ' start=' + res.timings.start + ' end=' + res.timings.end);
```

Comment (k-toumura) [reply packages/node_modules/@node-red/nodes/core/network/21-httprequest.js:376]: - We can also extend metric logs, such as:
```javascript
node.metric("duration.millis", msg, res.timings.phases.total);
node.metric("duration.wait.millis", msg, res.timings.phases.wait);
//... and others (dns, tcp, tls, request, firstByte, download)
```
- If httprequest node emit a message that contains entire `timings` object, following nodes can use this timing values in application (or use them for debugging purpose).
```javascript
msg.timings = res.timings
```

Comment (knolleary) [reply packages/node_modules/@node-red/nodes/core/network/21-httprequest.js:376]: I'm aware the current code doesn't yet do the metrics reporting the old node did. That is on my list to fix and yes, got makes it much easier to do then the old code.

But I'm not sure about adding timings to the message - we don't do that with any other node. I'm also not keen in exposing things the module gives us directly - because if we ever have to move off the got module, then it's even more work to reproduce the same behaviour.

As it stands, we're still blocked on this PR until we can get proxy support working. The issue i had raised on hpagent has not been answered - so need to look at a different plan for it.

Comment (kazuhitoyokoi) [reply packages/node_modules/@node-red/nodes/core/network/21-httprequest.js:376]: @knolleary @k-toumura Thank you for your comments. After supporting proxy in the new http request node, we can test it using our corporate proxy.

Comment (dceejay) [reply packages/node_modules/@node-red/nodes/core/network/21-httprequest.js:376]: could this be optional behind the metrics option in settings ?

Comment (knolleary) [reply packages/node_modules/@node-red/nodes/core/network/21-httprequest.js:376]: It could, but that doesn't change the fact it's becomes part of the public API of the node and puts us on the hook to ensure its always there

**Comment by hardillb** [comment]
Progress, I have proxying working but not if there is a redirect to SSL

Also it looks to be appending the the proxy port to the end of the URL in `msg.responseURL`

**Comment by hardillb** [comment]
One thing I have noticed is that the hpagent uses the CONNECT verb for all connection, so the proxy never gets to cache and some proxies might be configured not to support CONNECT for none HTTPS connections. Will need to remember to look out for that. There is a hpagent [bug](https://github.com/delvedor/hpagent/issues/17) asking for cache support (but it's been unanswered since Feb 2021).

File: packages/node_modules/@node-red/nodes/core/network/21-httprequest.js [hunk 2]
```diff
@@ -311,79 +336,96 @@ module.exports = function(RED) {
...
+                msg.headers = res.headers;
+                let hostname = res.req.host;
+                if (res.req.protocol === "http:" && res.socket.remotePort !== 80) {
```
[the remark this answers is not in the document]
Comment (hardillb) [reply packages/node_modules/@node-red/nodes/core/network/21-httprequest.js:389]: Fixed this in #2986

**Comment by hardillb** [comment]
I'm 95% sure the test failures are to do with the hpagent only using CONNECT to request content directly from the original target rather than from the cache

**Comment by knolleary** [comment]
Merging as-is - with more work on the proxy tests to follow

**Comment by kazuhitoyokoi** [comment]
http request node in the current dev branch returns HTTP 400 when using our corporate proxy. The first example using `http` module and `got` example in [the hpagent documentation](https://www.npmjs.com/package/hpagent ) also don't work in our environment. I will discuss it with my colleagues tomorrow.

**Comment by knolleary** [comment]
@kazuhitoyokoi thanks for letting us know. We'll go ahead with the beta release with a warning about the proxy support not being there yet. If there's debugging you can do to help track down a fix, that would be appreciated.

**Comment by hardillb** [comment]
When testing can you try a mix of http and https sites and also let us know how the proxy is setup:

 - Are you setting the env var or using the config node to set up the proxy?
 - Do you need to supply a username/password for the proxy?

Also is the proxy accessed via http or https? And if possible some wireshark/network capture logs might help

**Comment by kazuhitoyokoi** [comment]
@hardillb I submitted the issue (https://github.com/node-red/node-red/issues/3016) to answer all of the questions.
